stan/math/opencl/prim/cols.hpp

```diff
@@ -8,18 +8,19 @@
 namespace stan {
 namespace math {
 /** \ingroup opencl
- * Returns the number of columns in matrix_cl.
+ * Returns the number of columns in the specified kernel generator
+ * expression.
  *
- * @tparam T type of elements in the input matrix
- * @param x the input matrix_cl
+ * @tparam T_x type of input kernel generator expression x
+ * @param x input kernel generator expression.
  *
  * @return number of columns in x
  */
-template <typename T>
-inline int cols(const matrix_cl<T>& x) {
+template <typename T_x,
+          typename = require_all_kernel_expressions_and_none_scalar_t<T_x>>
+inline int cols(T_x&& x) {
```

Comment:
This function does not forward its argument. It should accept it as a const ref.
```suggestion
inline int cols(const T_x& x) {
```